File: source/lib/messaging/messaging.c
/* 
   Unix SMB/CIFS implementation.

   Samba internal messaging functions

   Copyright (C) Andrew Tridgell 2004
   
   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.
   
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.
   
   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

#include "includes.h"
#include "lib/events/events.h"
#include "system/filesys.h"
#include "system/time.h"
#include "messages.h"
#include "dlinklist.h"
#include "lib/socket/socket.h"

/* change the message version with any incompatible changes in the protocol */
#define MESSAGING_VERSION 1

/* maximum message size */
#define MESSAGING_MAX_SIZE 512

struct messaging_context {
	uint32_t server_id;
	struct socket_context *sock;
	const char *path;
	struct dispatch_fn *dispatch;
	struct messaging_rec *pending;

	struct {
		struct event_context *ev;
		struct fd_event *fde;
	} event;
};

/* we have a linked list of dispatch handlers that this messaging
   server can deal with */
struct dispatch_fn {
	struct dispatch_fn *next, *prev;
	uint32_t msg_type;
	void *private;
	void (*fn)(struct messaging_context *msg, void *private, 
		   uint32_t msg_type, uint32_t server_id, DATA_BLOB *data);
};

/* an individual message */
struct messaging_rec {
	struct messaging_rec *next, *prev;
	struct messaging_context *msg;
	const char *path;

	struct {
		uint32_t version;
		uint32_t msg_type;
		uint32_t from;
		uint32_t to;
		uint32_t length;
	} header;

	DATA_BLOB data;
};


/*
 A useful function for testing the message system.
*/
static void ping_message(struct messaging_context *msg, void *private, 
			 uint32_t msg_type, uint32_t src, DATA_BLOB *data)
{
	DEBUG(1,("INFO: Received PING message from server %u [%.*s]\n",
		 (uint_t)src, data->length, data->data?(const char *)data->data:""));
	messaging_send(msg, src, MSG_PONG, data);
}

/* 
   return the path to a messaging socket
*/
static char *messaging_path(TALLOC_CTX *mem_ctx, uint32_t server_id)
{
	char *name = talloc_asprintf(mem_ctx, "messaging/msg.%u", (unsigned)server_id);
	char *ret;
	ret = smbd_tmp_path(mem_ctx, name);
	talloc_free(name);
	return ret;
}

/*
  dispatch a fully received message
*/
static void messaging_dispatch(struct messaging_context *msg, struct messaging_rec *rec)
{
	struct dispatch_fn *d, *next;
	for (d=msg->dispatch;d;d=next) {
		next = d->next;
		if (d->msg_type == rec->header.msg_type) {
			d->fn(msg, d->private, d->msg_type, rec->header.from, &rec->data);
		}
	}
	rec->header.length = 0;
}


/*
  try to send the message
*/
static NTSTATUS try_send(struct messaging_rec *rec)
{
	struct messaging_context *msg = rec->msg;
	DATA_BLOB blob;
	size_t nsent;
	void *priv;
	NTSTATUS status;

	blob = data_blob_talloc(rec, NULL, sizeof(rec->header) + rec->data.length);
	NT_STATUS_HAVE_NO_MEMORY(blob.data);

	memcpy(blob.data, &rec->header, sizeof(rec->header));
	memcpy(blob.data + sizeof(rec->header), rec->data.data, rec->data.length);

	/* we send with privileges so messages work from any context */
	priv = root_privileges();
	status = socket_sendto(msg->sock, &blob, &nsent, 0, rec->path, 0);
	talloc_free(priv);

	data_blob_free(&blob);

	return status;
}

/*
  handle a socket write event
*/
static void messaging_send_handler(struct messaging_context *msg)
{
	while (msg->pending) {
		struct messaging_rec *rec = msg->pending;
		NTSTATUS status;
		status = try_send(rec);
		if (NT_STATUS_EQUAL(status, STATUS_MORE_ENTRIES)) {
			break;
		}
		if (!NT_STATUS_IS_OK(status)) {
			DEBUG(1,("messaging: Lost message from %u to %u of type %u - %s\n", 
				 rec->header.from, rec->header.to, rec->header.msg_type, 
				 nt_errstr(status)));
		}
		DLIST_REMOVE(msg->pending, rec);
		talloc_free(rec);
	}
	if (msg->pending == NULL) {
		EVENT_FD_NOT_WRITEABLE(msg->event.fde);
	}
}

/*
  handle a new incoming packet
*/
static void messaging_recv_handler(struct messaging_context *msg)
{
	struct messaging_rec *rec;
	NTSTATUS status;
	uint8_t data[MESSAGING_MAX_SIZE];
	size_t msize;

	status = socket_recv(msg->sock, data, sizeof(data), &msize, 0);
	if (!NT_STATUS_IS_OK(status)) {
		return;
	}

	if (msize < sizeof(rec->header)) {
		DEBUG(0,("messaging: bad message of size %d\n", msize));
		return;
	}

	rec = talloc(msg, struct messaging_rec);
	if (rec == NULL) {
		smb_panic("Unable to allocate messaging_rec");
	}

	rec->msg           = msg;
	rec->path          = msg->path;

	memcpy(&rec->header, data, sizeof(rec->header));
	if (msize != sizeof(rec->header) + rec->header.length) {
		DEBUG(0,("messaging: bad message header size %d should be %d\n", 
			 rec->header.length, msize - sizeof(rec->header)));
		talloc_free(rec);
		return;
	}

	rec->data = data_blob_talloc(rec, data+sizeof(rec->header), rec->header.length);
	if (rec->data.data == NULL) {
		talloc_free(rec);
		return;
	}

	messaging_dispatch(msg, rec);
	talloc_free(rec);
}


/*
  handle a socket event
*/
static void messaging_handler(struct event_context *ev, struct fd_event *fde, 
			      uint16_t flags, void *private)
{
	struct messaging_context *msg = talloc_get_type(private, 
							struct messaging_context);
	if (flags & EVENT_FD_WRITE) {
		messaging_send_handler(msg);
	}
	if (flags & EVENT_FD_READ) {
		messaging_recv_handler(msg);
	}
}


/*
  Register a dispatch function for a particular message type.
*/
void messaging_register(struct messaging_context *msg, void *private,
			uint32_t msg_type, 
			void (*fn)(struct messaging_context *, void *, uint32_t, uint32_t, DATA_BLOB *))
{
	struct dispatch_fn *d;

	d = talloc(msg, struct dispatch_fn);
	d->msg_type = msg_type;
	d->private = private;
	d->fn = fn;
	DLIST_ADD(msg->dispatch, d);
}

/*
  De-register the function for a particular message type.
*/
void messaging_deregister(struct messaging_context *msg, uint32_t msg_type, void *private)
{
	struct dispatch_fn *d, *next;

	for (d = msg->dispatch; d; d = next) {
		next = d->next;
		if (d->msg_type == msg_type && 
		    d->private == private) {
			DLIST_REMOVE(msg->dispatch, d);
			talloc_free(d);
		}
	}	
}


/*
  Send a message to a particular server
*/
NTSTATUS messaging_send(struct messaging_context *msg, uint32_t server, 
			uint32_t msg_type, DATA_BLOB *data)
{
	struct messaging_rec *rec;
	NTSTATUS status;

	rec = talloc(msg, struct messaging_rec);
	if (rec == NULL) {
		return NT_STATUS_NO_MEMORY;
	}

	rec->msg             = msg;
	rec->header.version  = MESSAGING_VERSION;
	rec->header.msg_type = msg_type;
	rec->header.from     = msg->server_id;
	rec->header.to       = server;
	rec->header.length   = data?data->length:0;
	if (rec->header.length != 0) {
		rec->data = data_blob_talloc(rec, data->data, data->length);
	} else {
		rec->data = data_blob(NULL, 0);
	}

	rec->path = messaging_path(rec, server);

	status = try_send(rec);
	if (NT_STATUS_EQUAL(status, STATUS_MORE_ENTRIES)) {
		if (msg->pending == NULL) {
			EVENT_FD_WRITEABLE(msg->event.fde);
		}
		DLIST_ADD(msg->pending, rec);
		return NT_STATUS_OK;
	}

	talloc_free(rec);

	return status;
}

/*
  Send a message to a particular server, with the message containing a single pointer
*/
NTSTATUS messaging_send_ptr(struct messaging_context *msg, uint32_t server, 
			    uint32_t msg_type, void *ptr)
{
	DATA_BLOB blob;

	blob.data = (void *)&ptr;
	blob.length = sizeof(void *);

	return messaging_send(msg, server, msg_type, &blob);
}


/*
  destroy the messaging context
*/
static int messaging_destructor(void *ptr)
{
	struct messaging_context *msg = ptr;
	unlink(msg->path);
	return 0;
}

/*
  create the listening socket and setup the dispatcher
*/
struct messaging_context *messaging_init(TALLOC_CTX *mem_ctx, uint32_t server_id, 
					 struct event_context *ev)
{
	struct messaging_context *msg;
	NTSTATUS status;
	char *path;

	msg = talloc(mem_ctx, struct messaging_context);
	if (msg == NULL) {
		return NULL;
	}

	/* create the messaging directory if needed */
	path = smbd_tmp_path(msg, "messaging");
	mkdir(path, 0700);
	talloc_free(path);

	msg->path = messaging_path(msg, server_id);
	msg->server_id = server_id;
	msg->dispatch = NULL;
	msg->pending = NULL;

	status = socket_create("unix", SOCKET_TYPE_DGRAM, &msg->sock, 0);
	if (!NT_STATUS_IS_OK(status)) {
		talloc_free(msg);
		return NULL;
	}

	/* by stealing here we ensure that the socket is cleaned up (and even 
	   deleted) on exit */
	talloc_steal(msg, msg->sock);

	status = socket_listen(msg->sock, msg->path, 0, 50, 0);
	if (!NT_STATUS_IS_OK(status)) {
		DEBUG(0,("Unable to setup messaging listener for '%s'\n", msg->path));
		talloc_free(msg);
		return NULL;
	}

	/* it needs to be non blocking for sends */
	set_blocking(socket_get_fd(msg->sock), False);

	msg->event.ev   = talloc_reference(msg, ev);
	msg->event.fde	= event_add_fd(ev, msg, socket_get_fd(msg->sock), 
				       EVENT_FD_READ, messaging_handler, msg);

	talloc_set_destructor(msg, messaging_destructor);
	
	messaging_register(msg, NULL, MSG_PING, ping_message);

	return msg;
}

File: python/samba/tests/messaging.py
# -*- coding: utf-8 -*-
#
# Unix SMB/CIFS implementation.
# Copyright © Jelmer Vernooij <jelmer@samba.org> 2008
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

"""Tests for samba.messaging."""
import samba
from samba.messaging import Messaging
from samba.tests import TestCase
import time
from samba.ndr import ndr_print
from samba.dcerpc import server_id
import random
import os
from samba.compat import integer_types


class MessagingTests(TestCase):

    def get_context(self, *args, **kwargs):
        kwargs['lp_ctx'] = samba.tests.env_loadparm()
        return Messaging(*args, **kwargs)

    def test_register(self):
        x = self.get_context()

        def callback():
            pass
        msg_type = x.register((callback, None))
        self.assertTrue(isinstance(msg_type, integer_types))
        x.deregister(callback, msg_type)

    def test_all_servers(self):
        x = self.get_context()
        self.assertTrue(isinstance(x.irpc_all_servers(), list))

    def test_by_name(self):
        x = self.get_context()
        for name in x.irpc_all_servers():
            self.assertTrue(isinstance(x.irpc_servers_byname(name.name), list))

    def test_unknown_name(self):
        x = self.get_context()
        self.assertRaises(KeyError,
                          x.irpc_servers_byname, "samba.messaging test NONEXISTING")

    def test_assign_server_id(self):
        x = self.get_context()
        self.assertTrue(isinstance(x.server_id, server_id.server_id))

    def test_add_remove_name(self):
        x = self.get_context()
        name = "samba.messaging test-%d" % random.randint(1, 1000000)
        x.irpc_add_name(name)
        name_list = x.irpc_servers_byname(name)
        self.assertEqual(len(name_list), 1)
        self.assertEqual(ndr_print(x.server_id),
                         ndr_print(name_list[0]))
        x.irpc_remove_name(name)
        self.assertRaises(KeyError,
                          x.irpc_servers_byname, name)

    def test_ping_speed(self):
        got_ping = {"count": 0}
        got_pong = {"count": 0}
        timeout = False

        msg_pong = 0
        msg_ping = 0

        server_ctx = self.get_context((0, 1))

        def ping_callback(got_ping, msg_type, src, data):
            got_ping["count"] += 1
            server_ctx.send(src, msg_pong, data)

        msg_ping = server_ctx.register((ping_callback, got_ping))

        def pong_callback(got_pong, msg_type, src, data):
            got_pong["count"] += 1

        client_ctx = self.get_context((0, 2))
        msg_pong = client_ctx.register((pong_callback, got_pong))

        # Try both server_id forms (structure and tuple)
        client_ctx.send((0, 1), msg_ping, "testing")

        client_ctx.send((0, 1), msg_ping, "testing2")

        start_time = time.time()

        # NOTE WELL: If debugging this with GDB, then the timeout will
        # fire while you are trying to understand it.

        while (got_ping["count"] < 2 or got_pong["count"] < 2) and not timeout:
            client_ctx.loop_once(0.1)
            server_ctx.loop_once(0.1)
            if time.time() - start_time > 1:
                timeout = True

        self.assertEqual(got_ping["count"], 2)
        self.assertEqual(got_pong["count"], 2)

    def test_pid_defaulting(self):
        got_ping = {"count": 0}
        got_pong = {"count": 0}
        timeout = False

        msg_pong = 0
        msg_ping = 0

        pid = os.getpid()
        server_ctx = self.get_context((pid, 1))

        def ping_callback(got_ping, msg_type, src, data):
            got_ping["count"] += 1
            server_ctx.send(src, msg_pong, data)

        msg_ping = server_ctx.register((ping_callback, got_ping))

        def pong_callback(got_pong, msg_type, src, data):
            got_pong["count"] += 1

        client_ctx = self.get_context((2,))
        msg_pong = client_ctx.register((pong_callback, got_pong))

        # Try one and two element tuple forms
        client_ctx.send((pid, 1), msg_ping, "testing")

        client_ctx.send((1,), msg_ping, "testing2")

        start_time = time.time()

        # NOTE WELL: If debugging this with GDB, then the timeout will
        # fire while you are trying to understand it.

        while (got_ping["count"] < 2 or got_pong["count"] < 2) and not timeout:
            client_ctx.loop_once(0.1)
            server_ctx.loop_once(0.1)
            if time.time() - start_time > 1:
                timeout = True

        self.assertEqual(got_ping["count"], 2)
        self.assertEqual(got_pong["count"], 2)
